ios/engine/KMEI/KeymanEngine/Classes/Manager.swift

```diff
@@ -613,191 +613,97 @@ public class Manager: NSObject, UIGestureRecognizerDelegate {
+        try FileManager.default.createDirectory(at: Storage.active.keyboardDir(forID: keyboardID),
+                                                withIntermediateDirectories: true)
+      } catch {
+        log.error("Could not create dir for download: \(error)")
+        throw KMPError.fileSystem
+      }
+
+      var haveInstalledOne = false
+      for keyboard in installableKeyboards {
+        let storedPath = Storage.active.keyboardURL(for: keyboard)
+
+        var installableFiles: [[Any]] = [["\(keyboardID).js", storedPath]]
+        if let osk = k.osk {
+          let oskPath = Storage.active.fontURL(forKeyboardID: keyboardID, filename: osk)
+          installableFiles.append([osk, oskPath])
+        }
+
+        if let font = k.font {
+          let displayPath = Storage.active.fontURL(forKeyboardID: keyboardID, filename: font)
+          installableFiles.append([font, displayPath])
+        }
+        do {
+          for item in installableFiles {
+            var filePath = folder
+            if(FileManager.default.fileExists(atPath: (item[1] as! URL).path)) {
+              try FileManager.default.removeItem(at: item[1] as! URL)
             }
+            filePath.appendPathComponent(item[0] as! String)
+            try FileManager.default.copyItem(at: filePath,
+                                             to: item[1] as! URL)
+
           }
+        } catch {
+          log.error("Error saving the download: \(error)")
+          throw KMPError.copyFiles
+        }
+        if !haveInstalledOne {
+          Manager.shared.addKeyboard(keyboard)
+          haveInstalledOne = true
         }
       }
-    } catch {
-      log.error("error parsing keyboard kmp: \(error)")
-      throw KMPError.invalidPackage
     }
   }
     
   // MARK: - Adhoc lexical models
   static public func parseLMKMP(_ folder: URL, isCustom: Bool) throws -> Void {
-    do {
-      var path = folder
-      path.appendPathComponent("kmp.json")
-      let data = try Data(contentsOf: path, options: .mappedIfSafe)
-      let jsonResult = try JSONSerialization.jsonObject(with: data, options: .mutableLeaves)
-      if let jsonResult = jsonResult as? [String:AnyObject] {
-        var version: String = "1.0"
-
-        if let info = jsonResult["info"] as? [String:AnyObject] {
-          if let versionEntry = info["version"] as? [String:AnyObject] {
-            if let description = versionEntry["description"] as? String {
-              version = description;
-            }
-          }
-        }
+    guard let kmp = KeymanPackage.parse(folder) as? LexicalModelKeymanPackage else {
+      throw KMPError.wrongPackageType
+    }
 
-        // Version uses a 'conditional initializer'.  If it fails, the version info is invalid.
-        if let _ = Version(version) {
-          // No problem
-        } else {
-          // Lazy-handle the error and replace version with 1.0.  Legacy decision from 2005.
-          version = "1.0"
-        }
+    for m in kmp.models {
```

Comment:
btw, there should only be one lexical model in a lexical model package

Reply:
While true for lexical models, it's not true for keyboards... and the next PR merges the control flow for both.  This sets up for that abstraction by highlighting how identical the two `parse___KMP` methods really are, even if one is technically more restricted than the other.